### ISSUE TITLE
Geocoding/search plugin/widget/control

### ISSUE THREAD

**Issue by prusswan** [developmentseed/lonboard#681]
**Is your feature request related to a problem? Please describe.**

Lonboard is great for visualizing large datasets, but sometimes I need to zero into a particular location so a search widget/control would be handy.

**Describe the solution you'd like**

Some means of building/integrating a search/geocoding control (that takes in textual input, turn it into location coordinates for the map to zoom/center to). User to provide JS event handling code for the respective search/geocoding service they prefer to use. But a sample can be provided for Nominatim/Mapbox etc

**Describe alternatives you've considered**

N/A. Might have to dig deeper into the widget internals if I had to do it myself

**Additional context**

I have some previous code done for folium (https://github.com/prusswan/folium/tree/multiple-and-generic-geocode-providers) - which is now able to support multiple geocoding services, although the custom geocode support (allowing users to define geocoding and event handling functions) was not incorporated due to maintenance reasons. (python-visualization/folium#1851)

**Comment by prusswan** [comment]
> Geocoding needs some sort of service running in the background, so it's unclear if we want that as part of the native Lonboard JS app. 

yeah for this to work as part of the exported html, there needs to be a way for the html to access an external service (could be anything from places autocomplete or some other search endpoint), I guess I am just looking for a least intrusive way to get this to work.

> You could make your own [anywidget](https://github.com/manzt/anywidget)-based widget with a little bit of JavaScript, and then wire up an event to set lonboard state.

A pure-Python widget should be sufficient for using within Jupyter. But if this can be part of the html, this will be more useful to the users who just want the visualization without having to deal with Python or JS. Actually, I just found #203 which is probably the other control I would need as well. Do you see any potential changes for lonboard to make it easier for users to add such controls (e.g. write less widget code)?

**Comment by kylebarron** [comment]
> A pure-Python widget should be sufficient for using within Jupyter. But if this can be part of the html, this will be more useful to the users who just want the visualization without having to deal with Python or JS.

Well, the ideal solution would be to have some sort of `lonboard-controls` library, that has its own controls that integrate with Lonboard via Python widget properties. Though this means the controls will be rendered next to the map instead of on the same HTML object.

**Comment by prusswan** [comment]
![ol_autocomplete_anywidget](https://github.com/user-attachments/assets/d9343766-2013-4352-9748-e3948a2feb70)


I managed to get something working using anywidgets. However, it is only working for one provider, since I don't have a good solution yet for accepting provider options (minimally I think at least two JS functions will be needed for every provider - one for sending the request, and one for formatting the result). Ideally this could also be extended to support server-side queries against local data layers.

**Comment by kylebarron** [comment]
Oh that looks great!

**Comment by kylebarron** [comment]
@prusswan given #924 I think we can say this is now in scope.

We can use https://github.com/maplibre/maplibre-gl-geocoder for Maplibre-driven maps (when deck.gl is injected into Maplibre) and https://deck.gl/docs/api-reference/widgets/geocoder-widget for deck.gl-driven maps, and we can follow the layout of existing controls defined in #924

Would you be interested in making a PR?

**Comment by kylebarron** [comment]
I started a draft in #991 but it's a little complex to use the maplibre-gl-geocoder API. 

If anyone is interested in this, help is welcome
